ports/espressif/common-hal/analogbufio/BufferedIn.c

```diff
@@ -0,0 +1,283 @@
+        mp_raise_ValueError_varg(translate("Unable to initialize ADC DMA controller, ErrorCode:%d"),err);
+    }
+
+    adc_digi_configuration_t dig_cfg = {
+        .conv_limit_en = ADC_CONV_LIMIT_EN,
+        .conv_limit_num = 250,
+        .pattern_num = NUM_ADC_CHANNELS,
+        .sample_freq_hz = sample_rate,
+        .conv_mode = convert_mode,
+        .format = output_format,
+    };
+
+#if defined(DEBUG_ANALOGBUFIO)
+    mp_printf(&mp_plat_print,"conversion_mode:%d, format:%d, conv_limit_en:%d, sample_rate:%d\n",convert_mode,output_format,ADC_CONV_LIMIT_EN,sample_rate);
+#endif //DEBUG_ANALOGBUFIO
+
+    adc_digi_pattern_config_t adc_pattern[NUM_ADC_CHANNELS] = {0};
+    adc_pattern[0].atten = ATTENUATION;
+    adc_pattern[0].channel = pin->adc_channel;
+    if(pin->adc_index == ADC_UNIT_1) {
+        adc_pattern[0].unit = 0;
+    } else {
+        adc_pattern[0].unit = 1;
+    }
+    adc_pattern[0].bit_width = SOC_ADC_DIGI_MAX_BITWIDTH;
+
+    dig_cfg.adc_pattern = adc_pattern;
+#if defined(DEBUG_ANALOGBUFIO)
+    mp_printf(&mp_plat_print,"adc_pattern[0].channel:%d, adc_pattern[0].unit:%d, adc_pattern[0].atten:%d\n",adc_pattern[0].channel,adc_pattern[0].unit,adc_pattern[0].atten);
+#endif //DEBUG_ANALOGBUFIO
+
+    err = adc_digi_controller_configure(&dig_cfg);
+    if(ESP_OK != err) {
+        common_hal_analogbufio_bufferedin_deinit(self);
+        mp_raise_ValueError_varg(translate("Unable to configure ADC DMA controller, ErrorCode:%d"),err);
+    }
+    err = adc_digi_start();
+    if(ESP_OK != err) {
+        common_hal_analogbufio_bufferedin_deinit(self);
+        mp_raise_ValueError_varg(translate("Unable to start ADC DMA controller, ErrorCode:%d"),err);
+    }
+}
+
+bool common_hal_analogbufio_bufferedin_deinited(analogbufio_bufferedin_obj_t *self) {
+    return self->pin == NULL;
+}
+
+void common_hal_analogbufio_bufferedin_deinit(analogbufio_bufferedin_obj_t *self) {
+    if (common_hal_analogbufio_bufferedin_deinited(self)) {
+        return;
+    }
+
+    adc_digi_stop();
+    adc_digi_deinitialize();
+
+    // Release ADC Pin
+    reset_pin_number(self->pin->number);
+    self->pin = NULL;
+}
+
+static bool check_valid_data(const adc_digi_output_data_t *data)
+{
+    unsigned int unit = data->type2.unit;
+    if(output_format == ADC_DIGI_OUTPUT_FORMAT_TYPE2) {
+        if (data->type2.channel >= SOC_ADC_CHANNEL_NUM(unit)) return false;
+        if (adc_channel != data->type2.channel) return false;
+    } else {
+        if( convert_mode == ADC_CONV_SINGLE_UNIT_1 ) {
+            unit = 0;
+        } else {
+            unit = 1;
+        }
+        if (data->type1.channel >= SOC_ADC_CHANNEL_NUM(unit)) return false;
+        if (adc_channel != data->type1.channel) return false;
+    }
+    if (unit > 2) return false;
+    return true;
+}
+
+
+uint32_t common_hal_analogbufio_bufferedin_readinto(analogbufio_bufferedin_obj_t *self, uint8_t *buffer, uint32_t len, uint8_t bytes_per_sample) {
```

Comment:
It looks like from the example that the allowed sample size varies per processor. It would be good to validate this by checking the element size of the array that is passed in, and raising errors if they don't fit. There is already a check for this in `shared-bindings/analogbufio/BufferedIn.c` in `...readinto()` there. Additional checks should be added here. It's a choice whether to make sure the sizes match perfectly, or allow 16 bit samples to be written to a 32-bit buffer.

Reply:
In `ports/espressif/esp-idf/components/hal/include/hal/adc_types.h` the structure `adc_digi_output_data_t` is used to return data read from DMA. The actual  ADC sample data is maximum up to 13 bits however it contains additional bits to identify channel and unit. Since for analogbufio we only return the sample values only 16-bit buffer would suffice. I will update the code accordingly and add checks to verify if application has provided 16-bit buffer.

Reply:
The code is updated as per above comment in commit https://github.com/adafruit/circuitpython/pull/7602/commits/ead5751803ae2c770ec5923c5c996cb1f33aa930

Reply:
I don't see any uses of `bytes_per_sample` at all besides the reference in the signature. Did you mean to check that it was `2`?

Reply:
Yes, the check is for printing the returned result from DMA engine which depends on ESP32 chip version it might be 2 byte result or 4 byte result depending on chip version.  The variable bytes_per_sample is provided by the application hence it is not used here.

Reply:
`bytes_per_sample` is set in the Python API interface, based on the whether the array passed in is bytes or halfwords: https://github.com/adafruit/circuitpython/blob/508f2015b68042c25256a8f57dd37e5aec2c48c5/shared-bindings/analogbufio/BufferedIn.c#L144-L149
I'm actually confused why your example in the top post works at all, because it looks like the validation in the lines above should reject the `'L'` `array.array`. Your `check_valid_data` does not throw an error if the size is a mismatch, it just prints something if the debugging output is on.

For any of the ESP32-nn chips, do you always want to return a 16-bit result? If so, then you could check that `bytes_per_sample == 2` and raise an error immediately if it is not. Or, if you want to allow one byte per sample, the code could shift the result to the right and drop some bits.

But you example would pass `bytes_per_sample == 4` for the `array.array('L', ...)`.

Perhaps the signature for `common_hal_analogbufio_bufferedin_readinto()` should have `uint8_t *buffer` be `void *buffer` instead, because the elements will vary in size based on what the original array is.

Reply:
The sample python code and the C code have been updated by later commits. The previous sample was only tested for ESP32S2 chip and not for other versions as noted in my original post. (Even the modified sample is used with ESP32S2 only). Hence `'L' ` `array.array` was never used. It would have thrown error if tested for other chips. Please refer to updated sample it no longer use` 'L' ` `array.array`

As per updated code the returned sample value is always 16-bit for ES32-nn. The application can pass 1-byte array or 16-bit element array, the result will always be copied in 16-bit values in provided buffer.  If the application wants to collect 16 samples it could provide `'B' 'array.array' `of 32 elements or `'H' array.array `of 16 elements that choice is left to application. Since allowed buffer types is `'B' `and `'H' `no further validity check required in addition to existing check.